pymatgen/io/input.py

```diff
@@ -0,0 +1,172 @@
+        are filenames, values are the contents of each file.
+
+        This method is called by write_input(), which performs the actual file
+        write operations.
+        """
+        pass
+
+    def write_input(
+        self,
+        directory: Union[str, Path],
+        make_dir: bool = True,
+        overwrite: bool = True,
+        zip_inputs: bool = False,
+        **kwargs,
+    ):
+        """
+        Write Inputs to one or more files
+
+        Args:
+            directory: Directory to write input files to
+            make_dir: Whether to create the directory if it does not already exist.
+            overwrite: Whether to overwrite an input file if it already exists.
+            Additional kwargs are passed to generate_inputs
+            zip_inputs: If True, inputs will be zipped into a file with the
+                same name as the InputSet (e.g., InputSet.zip)
+        """
+        path = directory if isinstance(directory, Path) else Path(directory)
+        # the following line will trigger a mypy error due to a bug in mypy
+        # will be fixed soon. See https://github.com/python/mypy/commit/ea7fed1b5e1965f949525e918aa98889fb59aebf
+        files = self._generate_input_data(**kwargs)  # type: ignore
+        for fname, contents in files.items():
+            file = path / fname
+
+            if not path.exists():
+                if make_dir:
+                    path.mkdir(parents=True, exist_ok=True)
+
+            if file.exists() and not overwrite:
+                raise FileExistsError(f"File {str(fname)} already exists!")
+            file.touch()
+
+            # write the file
+            with zopen(file, "wt") as f:
+                f.write(contents)
+
+        if zip_inputs:
+            zipfilename = path / f"{self.__class__.__name__}.zip"
+            with ZipFile(zipfilename, "w") as zip:
+                for fname, contents in files.items():
+                    file = path / fname
+                    try:
+                        zip.write(file)
+                        os.remove(file)
+                    except FileNotFoundError:
+                        pass
+
+    @classmethod
+    @abc.abstractmethod
+    def from_directory(cls, directory: Union[str, Path]):
+        """
+        Construct an InputSet from a directory of one or more files.
+
+        Args:
+            directory: Directory to read input files from
+        """
+        raise NotImplementedError(f"from_directory has not been implemented in {cls}")
+
+    def validate(self) -> bool:
+        """
+        A place to implement basic checks to verify the validity of an
+        input set. Can be as simple or as complex as desired.
+
+        Will raise a NotImplementedError unless overloaded by the inheriting class.
+        """
+        raise NotImplementedError(f".validate() has not been implemented in {self.__class__}")
+
+
+class InputSetGenerator(MSONable):
```

Comment:
Semantically, since `InputSetGenerator` is supposed to generate `InputSet`, don't you think the names should be consistent? I agree `InputSetGenerator` is a little long but I think the consistency between the two is helpful for clarity. Alternatively, we could rename `InputSet` to `Input`.

Reply:
No. InputSet should be a Set. So it should be an InputSet. But a generator can generate a single input or a SET of inputs. Some codes require a single input file, e.g., PWSCF.

Reply:
Why do we need both `InputSet` and `Input` classes? I think this proposal is to have `InputSet` represent **one or more** inputs with tools for writing and reading them to/from a directory.

Reply:
We do not need both. But there are times you handle a single input (e.g., INCAR) and there are times you handle a set of inputs (INCAR, POSCAR, POTCAR, KPOINTS). A generator should be agnostic to both.

Reply:
> Why do we need both InputSet and Input classes? I think this proposal is to have InputSet represent one or more inputs with tools for writing and reading them to/from a directory.

Yes, that's exactly the way I've been envisioning this